packages/core/src/browser/markdown-renderer.ts

```diff
@@ -0,0 +1,83 @@
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as DOMPurify from 'dompurify';
+import * as markdownit from 'markdown-it';
+
+export class MarkdownRenderer {
+
+    protected engine: markdownit;
+    protected domParser = new DOMParser();
+    protected callbacks = new Map<string, ((element: Element) => Element | void)[]>();
+
+    constructor(engine?: markdownit) {
+        this.engine = engine ?? markdownit();
+    }
+
+    /**
+     * Adds a modification callback that is applied to every element with the specified tag after rendering to HTML.
+     *
+     * @param tag The tag that this modification applies to
+     * @param callback The modification to apply on every selected rendered element. Can either modify the element in place or return a new element.
+     */
+    modify<K extends keyof HTMLElementTagNameMap>(tag: K, callback: (element: HTMLElementTagNameMap[K]) => Element | void): MarkdownRenderer {
+        if (this.callbacks.has(tag)) {
+            this.callbacks.get(tag)!.push(callback);
+        } else {
+            this.callbacks.set(tag, [callback]);
+        }
+        return this;
+    }
+
+    render(markdown: string): HTMLElement {
+        return this.renderInternal(this.engine.render(markdown));
+    }
+
+    renderInline(markdown: string): HTMLElement {
+        return this.renderInternal(this.engine.renderInline(markdown));
+    }
+
+    protected renderInternal(renderedHtml: string): HTMLElement {
+        const div = this.sanitizeHtml(renderedHtml);
+        for (const [tag, calls] of this.callbacks) {
+            // As the `taggedElements` collection may change while calling `replaceChild`,
+            // we have to create a copy of the collection beforehand.
+            const taggedElements = div.getElementsByTagName(tag);
+            const elements: Element[] = [];
+            for (let i = 0; i < taggedElements.length; i++) {
+                const element = taggedElements.item(i);
+                if (element) {
+                    elements.push(element);
+                }
+            }
+            for (const callback of calls) {
+                for (const element of elements) {
+                    const result = callback(element);
+                    if (result) {
+                        div.replaceChild(result, element);
+                    }
+                }
+            }
```

Comment:
This looks like it could easily fail. If there are multiple callbacks, and one or more of them return a new element in place of the original element, then the `div.replaceChild` call will subsequently fail because `element` is no longer a child of `div`: it will have been replaced by a previously-run callback.

This would be the scenario:

![image](https://user-images.githubusercontent.com/62660806/142686754-c4a65b0d-5145-44ff-97d5-8e6df51b5c63.png)

Reply:
I restructured this loop, as using multiple callbacks for a single `tag` was able to violate the contract. For example, if someone has two callbacks for `a` elements and the first callback builds `div` from them, the second callback would operate on these `div` elements, instead of re-selecting all `a` elements inside our rendered markdown.

This shouldn't fail anymore, as we re-select all tagged elements in each callback loop.